### PR TITLE
fix(scheduling queue): ignore events that interest no registered plugin

### PR DESCRIPTION
**Pull request by sanposhiho**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

ignore events that interest no registered plugin. It'll reduce the number of events stored in inFlightEvents.

#### Which issue(s) this PR fixes:

Part of #120622

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/scheduler/internal/queue/scheduling_queue_test.go [hunk 1]
```diff
@@ -223,33 +221,19 @@ func Test_InFlightPods(t *testing.T) {
 			},
 			wantInFlightPods:   nil,
 			wantInFlightEvents: nil,
-		},
-		{
-			name:        "when SchedulingQueueHint is disabled, which queue to enqueue Pod should be decided without SchedulingQueueHint",
```
Comment (sanposhiho) [inline pkg/scheduler/internal/queue/scheduling_queue_test.go:228]: Just moved to L391 with renaming to organize the order of test cases - tests of how we determine which queue to requeue are coming after tests of how we register inFlightPods/inFlightEvents.

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 2]
```diff
@@ -1141,6 +1161,11 @@ func (p *PriorityQueue) requeuePodViaQueueingHint(logger klog.Logger, pInfo *fra
 
 // NOTE: this function assumes lock has been acquired in caller
 func (p *PriorityQueue) movePodsToActiveOrBackoffQueue(logger klog.Logger, podInfoList []*framework.QueuedPodInfo, event framework.ClusterEvent, oldObj, newObj interface{}) {
+	if !p.isInterestedEvent(logger, event) {
```
Comment (AxeZhan) [inline pkg/scheduler/internal/queue/scheduling_queue.go:1164]: We can also add this check for  `moveAllToActiveOrBackoffQueue` to save one iteration for unschedulablePods?
https://github.com/kubernetes/kubernetes/blob/60dcf7fd8dbbb05fa02a08217001f98c5b1507c2/pkg/scheduler/internal/queue/scheduling_queue.go#L1081

**Comment by sanposhiho** [comment]
@kubernetes/sig-scheduling-approvers Can anyone take a look?

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 3]
```diff
@@ -409,6 +409,26 @@ const (
...
 
+// isInterestedEvent returns true if the event is interested by some plugins.
+func (p *PriorityQueue) isInterestedEvent(logger klog.Logger, event framework.ClusterEvent) bool {
+	if event.IsWildCard() {
+		return true
+	}
+
+	for _, hintMap := range p.queueingHintMap {
+		for eventToMatch := range hintMap {
+			if eventToMatch.Resource == event.Resource && eventToMatch.ActionType&event.ActionType != 0 {
```
[the remark this answers is not in the document]
Comment (kerthcet) [reply pkg/scheduler/internal/queue/scheduling_queue.go:420]: Created a relative PR here https://github.com/kubernetes/kubernetes/pull/123117.

**Review by kerthcet — COMMENTED** [review]
So here we're checking whether  a coming event will match one of all the plugin registered events here, I'm not doubt about the motivation with the PR, just wonder how ofter will this happen, because on my understanding, this is **rare** in the default scheduler. 🤔

**Comment by kerthcet** [comment]
> So here we're checking whether a coming event will match one of all the plugin registered events here, I'm not doubt about the motivation with the PR, just wonder how ofter will this happen, because on my understanding, this is rare in the default scheduler. 🤔

Can you help me to understand this? We have a bunch of events in  `queueingHintMap`, what we do here is to verify whether the coming event will match **one** event in the map, then most of the case, it will pass, but this will consume additional computing resources. Is this beneficial?

**Comment by sanposhiho** [comment]
@kerthcet 
As you said, it's little beneficial for the default scheduler at the moment.
But, this PR will be beneficial after https://github.com/kubernetes/kubernetes/pull/122628 - After #122628 many `AssignedPodUpdate` events can be dropped by this feature before being stored in `inFlightEvents`, because in-tree plugins register smaller Pod events (`PodRequestChange`/`PodLabelChange`) instead of `AssignedPodUpdate`.

**Comment by alculquicondor** [comment]
/approve
/hold
for rebase

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 4]
```diff
@@ -409,6 +409,26 @@ const (
...
 
+// isInterestedEvent returns true if the event is interested by some plugins.
+func (p *PriorityQueue) isInterestedEvent(logger klog.Logger, event framework.ClusterEvent) bool {
+	if event.IsWildCard() {
+		return true
+	}
+
+	for _, hintMap := range p.queueingHintMap {
+		for eventToMatch := range hintMap {
+			if eventToMatch.Match(event) {
```
Comment (sanposhiho) [inline pkg/scheduler/internal/queue/scheduling_queue.go:420]: Rebased upon the latest master. 
I only changed here to use a new `Match` function introduced in @kerthcet 's PR.
https://github.com/kubernetes/kubernetes/pull/123117

**Comment by kerthcet** [comment]
Got that, so after https://github.com/kubernetes/kubernetes/pull/122628 we can filter out pod update events like pod status update, but one day if we have all kinds of dedicated events and register most of them, this is less valuable because most of the events are matched. Make sense to me now for the default scheduler.

**Comment by kerthcet** [comment]
After https://github.com/kubernetes/kubernetes/pull/122628, we'll register PodRequestChange/PodLabelChange for in-tree plugins, then for incoming Pod updates, we'll have all these kinds of events:
- PodRequestChange
- PodLabelChange
- PodRequestChange & PodLabelChange
- Update

Since the `QueueingHintMap` owns these events, we'll match anyway. I think we should add new events like podAnnotationChange, podStatusChange in https://github.com/kubernetes/kubernetes/pull/122628, then we can get better performance.

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 5]
```diff
@@ -409,6 +409,26 @@ const (
 	queueImmediately
 )
 
+// isInterestedEvent returns true if the event is interested by some plugins.
+func (p *PriorityQueue) isInterestedEvent(logger klog.Logger, event framework.ClusterEvent) bool {
```
Comment (ahg-g) [inline pkg/scheduler/internal/queue/scheduling_queue.go:413]:
```suggestion
func (p *PriorityQueue) isEventOfInterest(logger klog.Logger, event framework.ClusterEvent) bool {
```

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 6]
```diff
@@ -409,6 +409,26 @@ const (
 	queueImmediately
 )
 
+// isInterestedEvent returns true if the event is interested by some plugins.
```
Comment (ahg-g) [inline pkg/scheduler/internal/queue/scheduling_queue.go:412]:
```suggestion
// isEventOfInterest returns true if the event is of interest by some plugins.
```

**Comment by sanposhiho** [comment]
@kerthcet 

> I think we should add new events like podAnnotationChange, podStatusChange in https://github.com/kubernetes/kubernetes/pull/122628, then we can get better performance.

Those may benefit out-of-tree plugins though, https://github.com/kubernetes/kubernetes/pull/122292 already supports all events that in-tree plugins need.

**Comment by sanposhiho** [comment]
/unhold 

- rebase done.
- two reviews done.

**Pull request merged**